src/profile-migrator/ephy-profile-migrator.c

```diff
@@ -698,6 +701,172 @@ migrate_app_desktop_file_categories (void)
+
+      xmlFree (content);
+      return;
+    }
+
+    this_language = "C";
+  } else
+    this_language = (char *)lang;
+
+  locales = g_get_language_names ();
+
+  for (i = 0; locales[i] && i < *best_match; i++) {
+    if (!strcmp (locales[i], this_language)) {
+      /* if we've already encountered a less accurate
+       * translation, then free it */
+      if (*value) xmlFree (*value);
+
+      *value = content;
+      *best_match = i;
+
+      break;
+    }
+  }
+
+  if (lang) xmlFree (lang);
+  if (*value != content) xmlFree (content);
+}
+
+static void
+parse_rdf_item (EphyBookmarksManager *manager,
+                xmlNodePtr            node)
+{
+  xmlChar *title = NULL;
+  int best_match_title = INT_MAX;
+  xmlChar *link = NULL;
+  int best_match_link = INT_MAX;
+  /* we consider that it's better to use a non-localized smart link than
+   * a localized link */
+  gboolean use_smartlink = FALSE;
+  xmlChar *subject = NULL;
+  GSequence *tags;
+  xmlNode *child;
+
+  child = node->children;
+
+  link = xmlGetProp (node, (xmlChar *)"about");
+
+  tags = g_sequence_new (g_free);
+  while (child != NULL) {
+    if (xmlStrEqual (child->name, (xmlChar *)"title")) {
+      parse_rdf_lang_tag (child, &title, &best_match_title);
+    } else if (xmlStrEqual (child->name, (xmlChar *)"link") &&
+               !use_smartlink) {
+      parse_rdf_lang_tag (child, &link, &best_match_link);
+    } else if (child->ns &&
+               xmlStrEqual (child->ns->prefix, (xmlChar *)"ephy") &&
+               xmlStrEqual (child->name, (xmlChar *)"smartlink")) {
+      if (!use_smartlink) {
+        use_smartlink = TRUE;
+        best_match_link = INT_MAX;
+      }
+
+      parse_rdf_lang_tag (child, &link, &best_match_link);
+    } else if (child->ns &&
+               xmlStrEqual (child->ns->prefix, (xmlChar *)"dc") &&
+               xmlStrEqual (child->name, (xmlChar *)"subject")) {
+      subject = xmlNodeGetContent (child);
+      if (subject)
+        g_sequence_prepend (tags, subject);
+    }
+
+    child = child->next;
+  }
+
+  if (link) {
+    EphyBookmark *bookmark;
+
+    g_sequence_sort (tags, (GCompareDataFunc)ephy_bookmark_tags_compare, NULL);
+    bookmark = ephy_bookmark_new ((char *)link, (char *)title, tags);
+    ephy_bookmarks_manager_add_bookmark (manager, bookmark);
+  } else {
+    g_sequence_free (tags);
+  }
+
+  xmlFree (title);
+  xmlFree (link);
+}
+
+static void
+migrate_bookmarks (void)
+{
+  EphyBookmarksManager *manager = EPHY_BOOKMARKS_MANAGER (g_object_new (EPHY_TYPE_BOOKMARKS_MANAGER, NULL));
+
+  const char *filename;
+  xmlDocPtr doc;
+  xmlNodePtr child;
+  xmlNodePtr root;
+
+  filename = g_build_filename (ephy_dot_dir (),
+                               EPHY_BOOKMARKS_FILE_RDF,
+                               NULL);
+
+  if (g_settings_get_boolean (EPHY_SETTINGS_LOCKDOWN,
+                              EPHY_PREFS_LOCKDOWN_BOOKMARK_EDITING))
+    return;
+
+  if (g_file_test (filename, G_FILE_TEST_EXISTS) == FALSE)
```

Comment:
I think it's fine here, but in general note that file existence tests are rarely a good idea, you should just try to open the file and move on if it fails. Otherwise you have a race condition between testing the file and actually opening it, which often leads to security issues.